pkg/cri/opts/spec_linux.go

```diff
@@ -672,6 +672,17 @@ func WithPodNamespaces(config *runtime.LinuxContainerSecurityContext, sandboxPid
 	if namespaces.GetPid() != runtime.NamespaceMode_CONTAINER {
 		opts = append(opts, oci.WithLinuxNamespace(runtimespec.LinuxNamespace{Type: runtimespec.PIDNamespace, Path: GetPIDNamespace(targetPid)}))
 	}
+
+	if namespaces.GetUsernsOptions() != nil {
```

Comment:
I am not sure that this field should be aligned with podSandbox. and uids/gids.
For your integration case, you didn't set the `UsernsOptions` so that case fails. 

The following patch runs in my local. And it pass.


```diff
commit c10ecf4a61cf56cd73ff869c0d1a3998d0c49178 (HEAD -> review-7679)
Author: Wei Fu <fuweid89@gmail.com>
Date:   Thu Dec 22 23:02:26 2022 +0800

    --wip-- [skip ci]

diff --git a/pkg/cri/opts/spec_linux.go b/pkg/cri/opts/spec_linux.go
index 767c9c2fc..51356e4e6 100644
--- a/pkg/cri/opts/spec_linux.go
+++ b/pkg/cri/opts/spec_linux.go
@@ -661,7 +661,7 @@ func WithSupplementalGroups(groups []int64) oci.SpecOpts {
 }

 // WithPodNamespaces sets the pod namespaces for the container
-func WithPodNamespaces(config *runtime.LinuxContainerSecurityContext, sandboxPid uint32, targetPid uint32, uids, gids []runtimespec.LinuxIDMapping) oci.SpecOpts {
+func WithPodNamespaces(config *runtime.LinuxContainerSecurityContext, sandboxConfig *runtime.PodSandboxConfig, sandboxPid uint32, targetPid uint32, uids, gids []runtimespec.LinuxIDMapping) oci.SpecOpts {
        namespaces := config.GetNamespaceOptions()

        opts := []oci.SpecOpts{
@@ -673,8 +673,8 @@ func WithPodNamespaces(config *runtime.LinuxContainerSecurityContext, sandboxPid
                opts = append(opts, oci.WithLinuxNamespace(runtimespec.LinuxNamespace{Type: runtimespec.PIDNamespace, Path: GetPIDNamespace(targetPid)}))
        }

-       if namespaces.GetUsernsOptions() != nil {
-               switch namespaces.GetUsernsOptions().GetMode() {
+       if usernsOpt := sandboxConfig.GetLinux().GetSecurityContext().GetNamespaceOptions().GetUsernsOptions(); usernsOpt != nil {
+               switch usernsOpt.GetMode() {
                case runtime.NamespaceMode_NODE:
                        // Nothing to do. Not adding userns field uses the node userns.
                case runtime.NamespaceMode_POD:
diff --git a/pkg/cri/sbserver/container_create.go b/pkg/cri/sbserver/container_create.go
index 953609f33..65abdfce5 100644
--- a/pkg/cri/sbserver/container_create.go
+++ b/pkg/cri/sbserver/container_create.go
@@ -188,7 +188,7 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
        log.G(ctx).Debugf("Container %q spec: %#+v", id, spew.NewFormatter(spec))

        // Grab any platform specific snapshotter opts.
-       sOpts := snapshotterOpts(c.config.ContainerdConfig.Snapshotter, config)
+       sOpts := snapshotterOpts(c.config.ContainerdConfig.Snapshotter, config, sandboxConfig.GetLinux().GetSecurityContext().GetNamespaceOptions())

        // Set snapshotter before any other options.
        opts := []containerd.NewContainerOpts{
diff --git a/pkg/cri/sbserver/container_create_linux.go b/pkg/cri/sbserver/container_create_linux.go
index 71e6af0a6..d6462b5b0 100644
--- a/pkg/cri/sbserver/container_create_linux.go
+++ b/pkg/cri/sbserver/container_create_linux.go
@@ -314,7 +314,7 @@ func (c *criService) containerSpec(
        specOpts = append(specOpts,
                customopts.WithOOMScoreAdj(config, c.config.RestrictOOMScoreAdj),
                // TODO: This is a hack to make this compile. We should move userns support to sbserver.
-               customopts.WithPodNamespaces(securityContext, sandboxPid, targetPid, nil, nil),
+               customopts.WithPodNamespaces(securityContext, sandboxConfig, sandboxPid, targetPid, nil, nil),
                customopts.WithSupplementalGroups(supplementalGroups),
                customopts.WithAnnotation(annotations.ContainerType, annotations.ContainerTypeContainer),
                customopts.WithAnnotation(annotations.SandboxID, sandboxID),
@@ -602,6 +602,6 @@ func generateUserString(username string, uid, gid *runtime.Int64Value) (string,
 }

 // snapshotterOpts returns any Linux specific snapshotter options for the rootfs snapshot
-func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig) []snapshots.Opt {
+func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig, nsOpts *runtime.NamespaceOption) []snapshots.Opt {
        return []snapshots.Opt{}
 }
diff --git a/pkg/cri/server/container_create.go b/pkg/cri/server/container_create.go
index ea4d5a02a..f5736848d 100644
--- a/pkg/cri/server/container_create.go
+++ b/pkg/cri/server/container_create.go
@@ -184,7 +184,7 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
        log.G(ctx).Debugf("Container %q spec: %#+v", id, spew.NewFormatter(spec))

        // Grab any platform specific snapshotter opts.
-       sOpts, err := snapshotterOpts(c.config.ContainerdConfig.Snapshotter, config)
+       sOpts, err := snapshotterOpts(c.config.ContainerdConfig.Snapshotter, config, sandboxConfig.GetLinux().GetSecurityContext().GetNamespaceOptions())
        if err != nil {
                return nil, err
        }
diff --git a/pkg/cri/server/container_create_linux.go b/pkg/cri/server/container_create_linux.go
index 5181c5f5e..74d188da3 100644
--- a/pkg/cri/server/container_create_linux.go
+++ b/pkg/cri/server/container_create_linux.go
@@ -311,14 +311,14 @@ func (c *criService) containerSpec(
                targetPid = status.Pid
        }

-       uids, gids, err := parseUsernsIDs(nsOpts.GetUsernsOptions())
+       uids, gids, err := parseUsernsIDs(sandboxConfig.GetLinux().GetSecurityContext().GetNamespaceOptions().GetUsernsOptions())
        if err != nil {
                return nil, fmt.Errorf("user namespace configuration: %w", err)
        }

        specOpts = append(specOpts,
                customopts.WithOOMScoreAdj(config, c.config.RestrictOOMScoreAdj),
-               customopts.WithPodNamespaces(securityContext, sandboxPid, targetPid, uids, gids),
+               customopts.WithPodNamespaces(securityContext, sandboxConfig, sandboxPid, targetPid, uids, gids),
                customopts.WithSupplementalGroups(supplementalGroups),
                customopts.WithAnnotation(annotations.ContainerType, annotations.ContainerTypeContainer),
                customopts.WithAnnotation(annotations.SandboxID, sandboxID),
@@ -606,7 +606,6 @@ func generateUserString(username string, uid, gid *runtime.Int64Value) (string,
 }

 // snapshotterOpts returns any Linux specific snapshotter options for the rootfs snapshot
-func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig) ([]snapshots.Opt, error) {
-       nsOpts := config.GetLinux().GetSecurityContext().GetNamespaceOptions()
+func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig, nsOpts *runtime.NamespaceOption) ([]snapshots.Opt, error) {
        return snapshotterRemapOpts(nsOpts)
 }
diff --git a/pkg/cri/server/container_create_other.go b/pkg/cri/server/container_create_other.go
index acab67c11..92262bc81 100644
--- a/pkg/cri/server/container_create_other.go
+++ b/pkg/cri/server/container_create_other.go
@@ -55,6 +55,6 @@ func (c *criService) containerSpecOpts(config *runtime.ContainerConfig, imageCon
 }

 // snapshotterOpts returns snapshotter options for the rootfs snapshot
-func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig) ([]snapshots.Opt, error) {
+func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig, nsOpts *runtime.NamespaceOption) ([]snapshots.Opt, error) {
        return []snapshots.Opt{}, nil
 }
diff --git a/pkg/cri/server/container_create_windows.go b/pkg/cri/server/container_create_windows.go
index e11466545..7fb78a97d 100644
--- a/pkg/cri/server/container_create_windows.go
+++ b/pkg/cri/server/container_create_windows.go
@@ -145,7 +145,7 @@ func (c *criService) containerSpecOpts(config *runtime.ContainerConfig, imageCon
 }

 // snapshotterOpts returns any Windows specific snapshotter options for the r/w layer
-func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig) ([]snapshots.Opt, error) {
+func snapshotterOpts(snapshotterName string, config *runtime.ContainerConfig, nsOpts *runtime.NamespaceOption) ([]snapshots.Opt, error) {
        var opts []snapshots.Opt

        switch snapshotterName {
diff --git a/script/test/cri-integration.sh b/script/test/cri-integration.sh
index 8d1a2a29a..7a4e81c81 100755
--- a/script/test/cri-integration.sh
+++ b/script/test/cri-integration.sh
@@ -60,6 +60,7 @@ ${CMD} --test.run="${FOCUS}" --test.v \
   --image-list="${TEST_IMAGE_LIST:-}" && test_exit_code=$? || test_exit_code=$?

 if [[ "$test_exit_code" -ne 0 ]]; then
+  cat "$REPORT_DIR/containerd.log"
   if [[ -e "$GITHUB_WORKSPACE" ]]; then
     mkdir -p "$GITHUB_WORKSPACE/report"
     mv "$REPORT_DIR/containerd.log" "$GITHUB_WORKSPACE/report"
```

Reply:
Thanks! With this diff, the integration test passes too: https://pastebin.com/zC8i8bu0

It seems the kubelet always set the containerConfig with the userns configuration, so IMHO it looks fine to use nsOpts (that comes from the containerConfig, not sandboxConfig) but maybe we need to tweak the integration framework to have some helpers to always set the options in both (as kubelet does)?

Reply:
Synced with @rata . The kubelet always uses pods namespace to set container config.

https://github.com/kubernetes/kubernetes/blob/442574f3a7321ff0f34dec9bae67f28651dbabd2/pkg/kubelet/kuberuntime/security_context.go#L28

I think the `integration` framework should be aligned with kubelet.

Reply:
> but maybe we need to tweak the integration framework to have some helpers to always set the options in both (as kubelet does)?

Sure. Let's do it in the followup~

Reply:
Sure, closing this now. Please re-open or let me know if I should create an issue for this